### PR TITLE
fix(daemon): session capability isolation, enforcement, and lifecycle

### LINKED ISSUE #4610: Fix session-scoped tool capability discovery and refresh propagation
## Summary

Complete session-aware tool discovery and cache invalidation for the daemon/proxy transport.

This follows [PR #4601](https://github.com/kaeawc/auto-mobile/pull/4601). A device session can be bound by a tool call, but subsequent `tools/list` requests in the default proxy topology may still use an unbound shared client. Capability changes also need to reach the external proxy cache, not only the loopback MCP server.

## Scope

- Route `tools/list` through the transport's bound device-session client after binding.
- Publish `notifications/tools/list_changed` through the existing daemon broadcaster when:
  - a call first binds or changes its `sessionUuid`; and
  - `ide/setSessionToolCapability` persists a change.
- Keep session bindings owned by the applicable server/transport lifecycle, so they cannot leak between server instances or survive a reused session ID.

## Acceptance criteria

- An agent's first device-aware call binds the session; a subsequent proxied `tools/list` advertises that session's enabled capabilities.
- A persisted capability enable or disable invalidates the connected proxy's cached tool list and the next list reflects the change.
- One MCP server/transport cannot observe another server's bound session.
- Add fast deterministic coverage for proxy-mode binding, initial refresh, post-update refresh, and lifecycle isolation.

## Validation

- Run focused daemon/proxy and MCP tools-list tests.
- Run the relevant TypeScript build and lint checks.

### LINKED ISSUE #4611: Enforce session tool capabilities across plan and socket execution paths
## Summary

Apply session capability enforcement at every execution boundary, not only direct MCP `tools/call` requests.

This follows [PR #4601](https://github.com/kaeawc/auto-mobile/pull/4601). The outer `executePlan` capability check does not constrain plan steps resolved through the internal plan path, and socket-local app-data operations can bypass the MCP call guard.

## Scope

- Thread the effective device session/profile into the canonical plan-internal invocation seam and reject disabled step tools.
- Enforce the same `app-data-interop` policy for socket-local `ide/setKeyValue`, `ide/removeKeyValue`, and `ide/clearKeyValueFile` operations.
- Keep the capability map and denial behavior as the single source of truth; do not duplicate independent allowlists.

## Acceptance criteria

- A profile that permits `executePlan` but disables a step's capability cannot execute that step.
- A disabled `app-data-interop` capability blocks the equivalent socket-local operations for the same device session.
- Enabled profiles retain current direct, plan, and socket behavior.
- Add deterministic tests covering allowed and denied paths for each boundary.

## Validation

- Run focused plan executor, socket server, and MCP capability tests.
- Run the relevant TypeScript build and lint checks.

### LINKED ISSUE #4673: Proxy binding: retire remembered session on a real release signal, not just a TTL backstop
## Context

Follow-up from PR [#4655](https://github.com/kaeawc/auto-mobile/pull/4655) (issue #4610), review finding on `src/daemon/daemonMcpProxy.ts`.

The P1 "stop replaying sessions after ordinary expiry" was fixed with a **replay TTL** (`DAEMON_BOUND_SESSION_REPLAY_TTL_MS`, 30 min, mirroring `SessionManager.SESSION_TIMEOUT_MS`): once that window elapses on the injected clock with no explicit-`sessionUuid` call refreshing the binding, the proxy stops replaying the remembered UUID on sessionless calls.

This is a **conservative heuristic**, not a precise signal. It was chosen because the daemon currently gives the proxy no way to learn a session was released:

- `getOrCreateSession()` in `ToolExecutionContext.ts` silently recreates a released session with no error/marker on the response, so "detect on the next call" is not feasible without new daemon-side wiring.
- The `DaemonNotification` wire type carries only `method` (no `sessionId`), so a daemon→proxy "session released" notification would be a cross-cutting wire change.

## Why the heuristic is safe (but imperfect)

- The binding is only replayed on **sessionless** calls; any explicit-`sessionUuid` call refreshes the timestamp.
- A session kept alive purely by daemon heartbeats beyond 30 min with no explicit-`sessionUuid` call from this proxy could have its binding retired early. Benign today (sessionless calls simply stop being rewritten), but not exact.

## Proposed robust fix (acceptance criteria)

- Add a daemon→proxy signal that a specific session was released (extend `DaemonNotification` to carry the `sessionId`, or return a released/recreated marker from the session-resolution path).
- The proxy invalidates `boundSessionUuid` exactly when its bound session is released, and the TTL becomes a backstop rather than the primary mechanism.
- Deterministic coverage with `FakeTimer` + the notification/marker seam.

Low priority — the TTL fix already closes the P1's core risk (never replaying a definitely-expired session).

### PR DESCRIPTION
**Pull request by kaeawc**
## Summary

Isolate and correctly manage each daemon-proxy socket's session-scoped MCP capability profile and session binding across its full lifecycle, while continuing to serialize tool execution by physical device or session. Expanded during review to cover the whole session-capability + lifecycle surface (subsumes #4611 and implements the release signal tracked in #4673).

## Linked Issues

Closes [#4610](https://github.com/kaeawc/auto-mobile/issues/4610)
Closes [#4611](https://github.com/kaeawc/auto-mobile/issues/4611)
Closes [#4673](https://github.com/kaeawc/auto-mobile/issues/4673)

## What changed

- **Real session-released signal (replaces the TTL guess).** All daemon session releases (heartbeat, idle, plan) fan out through one `SessionManager` choke point to a new `SessionReleaseBroadcaster`; the socket server pushes a `notifications/session/released` frame to connected proxies, and `DaemonMcpProxy` clears its remembered binding the moment the bound session is actually released. The 30-min replay TTL is now only a dropped-frame backstop.
- **Capability enforcement across every routing path (#4611).** The MCP path now derives a capability session from `deviceId` (closing the explicit-deviceId bypass); derived device-label sessions are enforced as the **union** of base + derived (a label can re-enable a tool the base narrowed away — deliberate product decision); and routing vs. capability session are un-conflated so nested navigation/replay keep their derived identity.
- **Plan-release binding teardown (#4611).** A released plan session no longer keeps enforcing its stale profile: an injected teardown seam (and the release broadcaster) clear the server-side `SessionToolBinding` on real release.
- **Transport seeding.** Explicit-session `ide/getNavigationGraph` reads route to their own client instead of repurposing the bound transport; `listTools`/`resources` bind the session so a reconnect mid-discovery stays session-scoped.
- **Binding lifecycle correctness.** Admitted requests keep their bound client on mid-flight disconnect; the replay lease refreshes on implicit forwarded calls; idle cleanup re-arms after a queued-forward timeout; and a pre-handler plan rejection preserves the (still-live) binding, leaving clearing to the real release signal.

## Validation

- [x] `bun run typecheck` reports no new errors
- [x] `bun run lint` clean (no new violations)
- [x] `bun test test/daemon/ test/server/ test/features/toolCapabilities/ test/plan/` — 2766+ pass, 0 fail
- [x] Each fix added a test that fails before / passes after
- [x] New tests use fakes and `FakeTimer`; no new dependency
- [x] Rebased onto latest `main`

## Follow-ups

- Union derived-session semantics is deliberately permissive; revisit if least-privilege is later preferred.

### CODE REVIEW

File: src/daemon/socketServer.ts [hunk 1]
```diff
@@ -584,96 +600,154 @@ export class UnixSocketServer {
...
     request: DaemonRequest,
     socketSessionId: string,
-    fn: (forwardKey: string) => Promise<T>
+    fn: (route: McpForwardRoute) => Promise<T>
   ): Promise<T> {
-    return this.runKeyedMcpForward(initialKey, async () => {
-      const currentKey = this.getMcpForwardKey(request, socketSessionId);
-      if (currentKey !== initialKey) {
+    return this.runKeyedMcpForward(initialRoute.executionKey, async () => {
+      const currentRoute = this.getMcpForwardRoute(request, socketSessionId);
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/socketServer.ts:713]: Fixed in 2a41ec132. The same-executionKey branch now runs with the admitted `initialRoute` (client + session); a genuine target change still re-resolves via the rekey recursion. A mid-flight disconnect that clears the binding can no longer downgrade an admitted request to the shared unbound client. Covered by `socketServerMcpForwardSerialization.test.ts` → "keeps the admitted bound client when the socket disconnects before a queued sessionless device call runs" (fails on the old `currentRoute` code, passes now).

File: src/daemon/daemonMcpProxy.ts [hunk 2]
```diff
@@ -874,6 +881,27 @@ export class DaemonMcpProxy {
...
 
+  private withBoundSessionUuid(args: Record<string, unknown>): Record<string, unknown> {
+    if (
+      !this.boundSessionUuid ||
+      (args.sessionUuid !== undefined &&
+        (typeof args.sessionUuid !== "string" || args.sessionUuid.trim().length > 0))
+    ) {
+      return args;
+    }
+    return { ...args, sessionUuid: this.boundSessionUuid };
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/daemonMcpProxy.ts:999]: Fixed in 2a41ec132 via a replay TTL (`DAEMON_BOUND_SESSION_REPLAY_TTL_MS`, 30 min, mirroring `SessionManager.SESSION_TIMEOUT_MS`): once the window elapses with no explicit-`sessionUuid` call refreshing the binding, the remembered UUID is retired, so a sessionless call is no longer rewritten to a released session. FakeTimer tests pin both sides of the window. The daemon exposes no session-released signal today (getOrCreateSession recreates silently; `DaemonNotification` carries no sessionId), so the exact-signal approach is tracked as follow-up #4673.

File: src/daemon/socketServer.ts [hunk 3]
```diff
@@ -584,96 +600,154 @@ export class UnixSocketServer {
...
+        }
       }
     });
     return run;
   }
 
-  private runMcpForwardForCurrentKey<T>(
-    initialKey: string,
+  private async runWithActiveMcpClient<T>(clientKey: string, fn: () => Promise<T>): Promise<T> {
+    this.clearMcpClientIdleTimer(clientKey);
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/socketServer.ts:682]: Fixed in 2a41ec132. Idle-close scheduling moved into `runWithActiveMcpClient`'s `finally` (fires when the client's active-forward count reaches zero), symmetric with the clear-on-entry; the redundant schedule in the forward-body `finally` was removed. A forward that throws before its own cleanup (e.g. the pre-forward queue-timeout deadline) now still re-arms the idle close. FakeTimer test: "re-arms the idle close after a queued forward times out in the queue".

File: test/daemon/socketServerInputTypeText.test.ts [hunk 4]
```diff
@@ -564,6 +564,109 @@ describe("UnixSocketServer input/typeText", () => {
...
+    await flushMicrotasks();
+    releaseAppend();
+
+    expect((await firstAppend).success).toBe(true);
+    expect((await queuedMcpForward).success).toBe(true);
+
+    await fakeTimer.advanceTimeAsync(5 * 60 * 1000);
+
+    expect((await append("B")).success).toBe(true);
+    expect(factoryCalls).toBe(2);
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply test/daemon/socketServerInputTypeText.test.ts:673]: Fixed in 2a41ec132 — the test now asserts `factoryCalls === 1` before advancing the idle window (proving the helper was cached exactly once, not never-cached / immediately-evicted), then advances and re-appends to expect the rebuild.

File: src/daemon/daemonMcpProxy.ts [hunk 5]
```diff
@@ -861,7 +870,10 @@ export class DaemonMcpProxy {
     args: Record<string, unknown>
   ): Promise<any> {
     try {
-      return await this.withRecoverableReconnect(() => this.client!.callTool(name, args));
+      const forwardedArgs = this.withBoundSessionUuid(args);
+      const result = await this.withRecoverableReconnect(() => this.client!.callTool(name, forwardedArgs));
+      this.rememberSessionUuid(name, args);
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/daemonMcpProxy.ts]: Fixed in 2df28a67b. `callTool` now refreshes the replay lease off the **forwarded** args (`rememberSessionUuid(name, forwardedArgs)`), so a successful implicit sessionless call that carried the bound UUID and extended the live daemon session also renews `boundSessionUuidAt` — continuous implicit activity no longer looks idle to the TTL. Regression test: `daemonMcpProxy.test.ts` → "keeps replaying a bound session across continuous implicit activity beyond one TTL" (two implicit calls each within the TTL but cumulatively past it; fails on the old `args` code, passes now).

File: src/daemon/daemonMcpProxy.ts [hunk 6]
```diff
@@ -861,7 +870,10 @@ export class DaemonMcpProxy {
     args: Record<string, unknown>
   ): Promise<any> {
     try {
-      return await this.withRecoverableReconnect(() => this.client!.callTool(name, args));
+      const forwardedArgs = this.withBoundSessionUuid(args);
+      const result = await this.withRecoverableReconnect(() => this.client!.callTool(name, forwardedArgs));
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/daemonMcpProxy.ts:944]: Fixed in a223221b2. `callTool`'s catch now clears the bound session when `name === "executePlan"`, so a rejected plan (whose daemon-side release still runs in the lifecycle finally) no longer leaves a stale UUID to be replayed. Regression test: `daemonMcpProxy.test.ts` → "clears the binding when an executePlan rejects, so its released session is not resurrected" (fails without the clear, passes now).

File: src/daemon/daemonMcpProxy.ts [hunk 7]
```diff
@@ -874,6 +886,50 @@ export class DaemonMcpProxy {
...
+
+  private clearBoundSessionUuid(): void {
+    this.boundSessionUuid = undefined;
+    this.boundSessionUuidAt = undefined;
+  }
+
+  private rememberSessionUuid(name: string, args: Record<string, unknown>): void {
+    if (name === "executePlan") {
+      this.clearBoundSessionUuid();
+      return;
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/daemonMcpProxy.ts:1054]: Acknowledged and deferred to [#4611](https://github.com/kaeawc/auto-mobile/issues/4611) (tracked there just now). This is the plan-release-lifecycle race: the coherent fix requires `DefaultPlanLifecycleManager.afterExecution()` to await/report `releaseSession()` so the proxy clears its binding only on real release completion — daemon-side lifecycle work that is #4611's scoped remaining item ("clear direct MCP bindings after a plan releases its session"), and fundamentally needs the missing release signal tracked in [#4673](https://github.com/kaeawc/auto-mobile/issues/4673). The replay TTL bounds the exposure until then. Keeping it out of #4655 to avoid a partial/incoherent lifecycle change here.

File: src/daemon/constants.ts [hunk 8]
```diff
@@ -129,6 +129,23 @@ export const READINESS_PROBE_BACKOFF_MS = 150;
...
+/**
+ * How long the proxy will keep replaying a remembered session binding on
+ * sessionless calls before treating it as retired (issue #4610). It mirrors the
+ * daemon's session idle timeout (`SessionManager.SESSION_TIMEOUT_MS`, 30 min):
+ * once this window elapses with no forwarded call (explicit or implicit)
+ * refreshing the binding, the daemon session has certainly idle/heartbeat-expired, and
+ * replaying its UUID would silently recreate the released session and reacquire
+ * a device without the caller asking for it.
+ */
+export const DAEMON_BOUND_SESSION_REPLAY_TTL_MS = 30 * 60 * 1000;
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/constants.ts:147]: Fixed at the root in ef60db8 (commit ①): the daemon now emits a real `notifications/session/released` signal from the single `SessionManager` release choke point (covers heartbeat, idle, and plan releases), and the proxy clears its binding the moment the bound session is actually released. The 30-min TTL is now only a dropped-frame backstop, so a heartbeat-expired session is no longer replayed. See `src/server/sessionReleaseBroadcast.ts`, `daemon.ts`, `daemonMcpProxy.ts handleDaemonNotification`.

File: src/daemon/daemonMcpProxy.ts [hunk 9]
```diff
@@ -861,7 +870,16 @@ export class DaemonMcpProxy {
     args: Record<string, unknown>
   ): Promise<any> {
     try {
-      return await this.withRecoverableReconnect(() => this.client!.callTool(name, args));
+      const forwardedArgs = this.withBoundSessionUuid(args);
+      const result = await this.withRecoverableReconnect(() => this.client!.callTool(name, forwardedArgs));
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/daemonMcpProxy.ts]: Fixed in 748cec812 (commit ④) + fa40f42cc: `listTools` (and `resources/list`/`resources/list-templates`) now forward `withBoundSessionUuid({})`, so a recoverable reconnect mid-discovery re-seeds the fresh transport with the bound session and returns the session-scoped list instead of the unscoped surface. Test: `daemonMcpProxy.test.ts` listTools-after-reconnect carries the bound sessionUuid.

File: src/daemon/socketServer.ts [hunk 10]
```diff
@@ -584,96 +602,168 @@ export class UnixSocketServer {
...
     }
 
     if (request.method === "ide/getNavigationGraph") {
-      return this.getRequestArgumentScopeKey(request.params)
-        ?? this.getBoundMcpClientKey(socketSessionId)
-        ?? `method:${request.method}`;
+      const executionKey = this.getRequestArgumentScopeKey(request.params);
+      if (executionKey) {
+        if (boundRoute) {
+          return { ...boundRoute, executionKey };
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/socketServer.ts]: Fixed in 748cec812 (commit ④): an explicit-`sessionUuid` `ide/getNavigationGraph` read now routes to its OWN session-specific client (`socket:<sid>:session:<uuid>`) via the shared `sessionScopedForwardRoute` helper, instead of repurposing the socket's bound client A. A's transport is no longer silently rebound to B. Test: `socketServerIdeForwardRoute.test.ts`.

File: src/daemon/daemonMcpProxy.ts [hunk 11]
```diff
@@ -861,8 +870,25 @@ export class DaemonMcpProxy {
...
+      // Remember what was actually forwarded, not the caller's raw args. An
+      // implicit sessionless call injects the bound UUID into forwardedArgs and
+      // extends the live daemon session in getOrCreateSession(); refreshing the
+      // replay lease off forwardedArgs keeps continuous implicit activity from
+      // being mistaken for idleness, so a later reconnect re-seeds the still-live
+      // session instead of creating an unseeded transport (issue #4610).
+      this.rememberSessionUuid(name, forwardedArgs);
+      return result;
     } catch (error) {
+      if (name === "executePlan") {
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/daemonMcpProxy.ts]: Correct — fixed in 6f112f235. The blanket clear-on-executePlan-reject is removed: a pre-handler rejection (capability/schema failure in `index.ts`) leaves the daemon session LIVE, so the proxy no longer forgets it. The binding is now cleared authoritatively by the daemon's real session-released signal (commit ①) only on actual release. Test flipped to assert the binding is PRESERVED across a plan rejection.

File: src/server/toolRegistry.ts [hunk 12]
```diff
@@ -857,14 +931,28 @@ export class ToolRegistryClass {
...
+        const capabilitySessionManager = DaemonState.getInstance().isInitialized()
+          ? DaemonState.getInstance().getSessionManager()
+          : undefined;
+        const capabilityBaseSessionUuid = resolvedTarget.baseSessionUuid
+          ?? resolveCapabilityBaseSessionUuid(capabilityDerivedSessionUuid, capabilitySessionManager);
+        await assertToolEnabledForAnySession(
           name,
-          capabilitySessionUuid,
+          [capabilityBaseSessionUuid, capabilityDerivedSessionUuid],
           getToolCapabilityContext()?.sessionToolProfileService,
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/toolRegistry.ts]: Fixed in 80b073ddf. The public `tools/call` boundary in `src/server/index.ts` now derives the `${base}:${label}` candidate via a read-only `getDeviceLabelMap(sessionUuid)?.[label]` lookup (no device allocation) and calls `assertToolEnabledForAnySession([base, derived])`, so a base-disabled/derived-enabled tool is no longer rejected before label resolution. Non-labeled calls collapse to `[base]`, preserving prior behavior and tools/list filtering. New tests drive the real MCP fixture through the public gate.

File: src/server/toolRegistry.ts [hunk 13]
```diff
@@ -1001,7 +1090,7 @@ export class ToolRegistryClass {
         progress,
         signal,
         options.targetDevice,
-        invocation.sessionUuid,
+        invocation.routingSessionUuid,
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/toolRegistry.ts:1088]: Fixed in 80b073ddf. `invokeInternalTool`'s single-session pre-gate became a union via a new `assertToolEnabledUnion` helper that resolves the base from the derived routing session and asserts `[base, derived]` — and it now also covers the `targetDevice` path (which invokes the device-aware handler directly, bypassing the wrapper's own gate). Tests exercise a labeled base-enabled/derived-disabled step through `callInternal` and the targetDevice path.

File: test/daemon/daemonSessionReleaseSignal.test.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+    try {
+      await sessionManager.createSession("session-release-signal", "emulator-5554", "android");
+      await sessionManager.releaseSession("session-release-signal");
+
+      expect(emitted).toEqual(["session-release-signal"]);
+    } finally {
+      unsubscribe();
+      sessionManager.stopCleanupTimer();
+    }
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply test/daemon/daemonSessionReleaseSignal.test.ts]: Fixed in cc07827da. Both tests now inject a migrated in-memory DB (`createTestDatabase()`) into the SessionManager via the Daemon/SessionManager `DeviceSessionRepository` seam, so createSession/releaseSession persist without resolving the real `~/.auto-mobile` file DB (CLAUDE.md guard #3067).

File: src/daemon/socketServer.ts [hunk 15]
```diff
@@ -584,96 +640,194 @@ export class UnixSocketServer {
...
       }
-      return await fn(currentKey);
+      // The execution target is unchanged, so this request keeps the client and
+      // session it was admitted with. Re-resolving may replace a session-specific
+      // clientKey with the shared unbound client (e.g. a mid-flight disconnect
+      // cleared the binding before this recompute) under the same executionKey;
+      // that would run the admitted tool with no capability profile. Only the
+      // execution target may be re-resolved, never the admitted client/session
+      // (issue #4610).
+      return await this.runWithActiveMcpClient(initialRoute.clientKey, () => fn(initialRoute));
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/socketServer.ts:748]: Fixed in fce7dde14. In `runMcpForwardForCurrentRoute`'s unchanged-target branch, a queued route seeded for a specific daemon session that was RELEASED while queued now re-resolves (invokes `currentRoute`, unseeded) instead of replaying the released UUID. The discriminator is `!hasActiveDaemonSession(initialRoute.sessionUuid)` — a live session means a mid-flight disconnect (keep the admitted client, preserving the capability profile); a gone session means a real release (re-resolve). The existing disconnect test stays green, proving the profile-preservation guarantee is intact.

File: src/daemon/daemonMcpProxy.ts [hunk 16]
```diff
@@ -861,8 +893,25 @@ export class DaemonMcpProxy {
...
-      return await this.withRecoverableReconnect(() => this.client!.callTool(name, args));
+      const forwardedArgs = this.withBoundSessionUuid(args);
+      const result = await this.withRecoverableReconnect(() => this.client!.callTool(name, forwardedArgs));
+      // Remember what was actually forwarded, not the caller's raw args. An
+      // implicit sessionless call injects the bound UUID into forwardedArgs and
+      // extends the live daemon session in getOrCreateSession(); refreshing the
+      // replay lease off forwardedArgs keeps continuous implicit activity from
+      // being mistaken for idleness, so a later reconnect re-seeds the still-live
+      // session instead of creating an unseeded transport (issue #4610).
+      this.rememberSessionUuid(name, forwardedArgs);
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/daemonMcpProxy.ts]: Fixed in fce7dde14. On an admitted-then-rejected implicit bound call, `callTool`'s catch now refreshes the replay lease (`refreshReplayLeaseAfterAdmittedFailure`) when the bound UUID rode along in forwardedArgs and the error is a genuine handler rejection — excluding `executePlan` (owns its binding via the release signal) and `isRecoverableDaemonSessionError` transport/connect/'Session not found' cases that never reached a live handler. So repeated admitted failures no longer let the daemon session outlive the proxy lease. A false-positive refresh (pre-handler rejection) is self-correcting: the real session-released signal clears the binding on actual release.

File: src/daemon/socketServer.ts [hunk 17]
```diff
@@ -584,96 +640,194 @@ export class UnixSocketServer {
...
+  private getMcpForwardRoute(request: DaemonRequest, socketSessionId: string): McpForwardRoute {
     if (request.method === "tools/call") {
-      return this.getToolsCallForwardKey(request.params?.arguments, socketSessionId);
+      return this.getToolsCallForwardRoute(request.params?.arguments, socketSessionId);
     }
 
+    const boundRoute = this.getBoundMcpClientRoute(socketSessionId);
     if (request.method === "tools/list") {
-      return this.getBoundMcpClientKey(socketSessionId) ?? `method:${request.method}`;
+      return boundRoute ?? this.sharedMcpForwardRoute(`method:${request.method}`);
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/socketServer.ts:769]: Fixed in fce7dde14. The `tools/list` route branch now honors `getSessionUuid(request.params)` and routes to `sessionScopedForwardRoute`, so a reconnected restricted discovery seeds the fresh loopback transport and returns the session-scoped list instead of the shared unseeded client. This completes the proxy-side seeding merged earlier. Test: `socketServerMcpReconnect.test.ts` (clientBindings === ["session-a"], scoped list).

File: src/server/index.ts [hunk 18]
```diff
@@ -317,7 +349,28 @@ export const createMcpServer = (options: McpServerOptions = {}): McpServer => {
...
+    const requestedDeviceLabel = typeof (toolParams as Record<string, unknown>).device === "string"
+      ? (toolParams as Record<string, unknown>).device as string
+      : undefined;
+    const derivedLabelSessionUuid = requestedDeviceLabel && sessionUuid
+      ? getDeviceLabelMap(sessionUuid)?.[requestedDeviceLabel]
+      : undefined;
+    await assertToolEnabledForAnySession(
+      name,
+      [sessionUuid, derivedLabelSessionUuid],
+      options.sessionToolProfileService,
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/index.ts:401]: Fixed in fce7dde14. The `ListToolsRequestSchema` handler now unions over `[base, ...Object.values(getDeviceLabelMap(base))]` via `isToolEnabledForAnySession`, mirroring the call-gate — so a tool a mapped label re-enables is now advertised in discovery, consistent with the union product decision. Test through the public MCP fixture: base-disables/label-enables now advertises the tool; neither-enables does not.

File: src/server/index.ts [hunk 19]
```diff
@@ -317,7 +360,28 @@ export const createMcpServer = (options: McpServerOptions = {}): McpServer => {
...
+    // gate could observe that `base:label` re-enables it. Resolve the derived
+    // label candidate from the base session's label map (a read-only lookup, no
+    // device allocation). Non-labeled and non-device-aware calls collapse to the
+    // base, preserving prior single-session behavior and `tools/list` filtering.
+    const requestedDeviceLabel = typeof (toolParams as Record<string, unknown>).device === "string"
+      ? (toolParams as Record<string, unknown>).device as string
+      : undefined;
+    const derivedLabelSessionUuid = requestedDeviceLabel && sessionUuid
+      ? getDeviceLabelMap(sessionUuid)?.[requestedDeviceLabel]
+      : undefined;
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/index.ts]: Fixed in 2dd46b228. The derived device-label candidate now joins the union only when `tool.requiresDevice` (set by `registerDeviceAware`; `register` sets it false). At the two MCP boundary surfaces where the `device` field is stripped by a plain tool's schema — the `tools/call` gate and the `tools/list` filter — a plain tool is enforced/advertised base-only, so `{sessionUuid: base, device: B}` can no longer borrow B's grants for a tool that runs under base. The nested/device-aware gates keep the union deliberately: there the `base:label` routing session is the genuine execution session, not a stripped device field. Tests: plain `exportPlan` base-disabled/label-enabled → rejected + not advertised; device-aware control still unioned.

File: src/daemon/daemonMcpProxy.ts [hunk 20]
```diff
@@ -860,9 +892,33 @@ export class DaemonMcpProxy {
...
     try {
-      return await this.withRecoverableReconnect(() => this.client!.callTool(name, args));
+      const result = await this.withRecoverableReconnect(() => this.client!.callTool(name, forwardedArgs));
+      // Remember what was actually forwarded, not the caller's raw args. An
+      // implicit sessionless call injects the bound UUID into forwardedArgs and
+      // extends the live daemon session in getOrCreateSession(); refreshing the
+      // replay lease off forwardedArgs keeps continuous implicit activity from
+      // being mistaken for idleness, so a later reconnect re-seeds the still-live
+      // session instead of creating an unseeded transport (issue #4610).
+      this.rememberSessionUuid(name, forwardedArgs);
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/daemonMcpProxy.ts]: Fixed in 2dd46b228. A monotonic `bindingGeneration` guards the in-flight race: `callTool` snapshots it at forward time, `clearBoundSessionUuid` (the single choke point reached by the release signal, close, and replay-expiry) bumps it, and both `rememberSessionUuid` and `refreshReplayLeaseAfterAdmittedFailure` skip re-establishing the UUID when the generation changed during the call. So a release observed mid-call is preserved — the next sessionless call is not rewritten to the released UUID — for both fulfilled and admitted-then-rejected calls; the no-release path re-remembers normally.

File: src/daemon/daemonMcpProxy.ts [hunk 21]
```diff
@@ -874,6 +942,104 @@ export class DaemonMcpProxy {
...
+      return;
+    }
+    // A release for the bound UUID observed WHILE this call was in flight already
+    // cleared the binding (handleDaemonNotification). Re-remembering the forwarded
+    // UUID now would resurrect the freed session and let the next sessionless call
+    // recreate it (issue #4611). The generation advancing since forward time is the
+    // signal; a later explicit call re-binds normally.
+    if (this.bindingGeneration !== callBindingGeneration) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/daemonMcpProxy.ts]: Fixed in 5764262af. The global `bindingGeneration` was replaced with a per-UUID release record (`releaseEpoch` + `releasedSessionEpochs: Map<uuid,epoch>`), recorded for every session-released frame. `callTool` captures the epoch at forward time and remember/refresh now ask `wasForwardedSessionReleasedSince(forwardedArgs, epoch)` — true only when the SPECIFIC forwarded UUID was released. So an unrelated session (A) expiring while an explicit B call is in flight no longer blocks remembering B, while a release of the forwarded UUID mid-call is still preserved. (Unbounded-Map pruning tracked as a hardening follow-up: #FU_PLACEHOLDER.)

File: src/server/toolRegistry.ts [hunk 22]
```diff
@@ -1033,13 +1120,46 @@ export class ToolRegistryClass {
...
+    toolName: string,
+    derivedSessionUuid: string | undefined,
+    explicitBaseSessionUuid: string | undefined,
+    sessionToolProfileService: Pick<SessionToolProfileService, "isEnabled"> | undefined,
+  ): Promise<void> {
+    const sessionManager = DaemonState.getInstance().isInitialized()
+      ? DaemonState.getInstance().getSessionManager()
+      : undefined;
+    const baseSessionUuid = explicitBaseSessionUuid
+      ?? resolveCapabilityBaseSessionUuid(derivedSessionUuid, sessionManager);
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/toolRegistry.ts]: Fixed in 5764262af. `assertToolEnabledUnion` now resolves the base unconditionally via `resolveCapabilityBaseSessionUuid(explicitBaseSessionUuid ?? derivedSessionUuid, sessionManager)`, so the wrapper no longer short-circuits on the already-derived `base:B` — the union is genuinely `[base, base:B]` and a base-enabled/derived-disabled internal step is allowed. Test through `callInternal` with DaemonState labels set.

File: src/daemon/daemonMcpProxy.ts [hunk 23]
```diff
@@ -841,8 +874,14 @@ export class DaemonMcpProxy {
...
+      // with empty params against the fresh UNSEEDED transport, which returns the
+      // full unfiltered tool list instead of the session-scoped one. Computing the
+      // forwarded params inside the operation closure re-seeds the retry after a
+      // reconnect (issue #4610).
       const result = await this.withRecoverableReconnect(() =>
-        this.client!.callDaemonMethod("tools/list", {})
+        this.client!.callDaemonMethod("tools/list", this.withBoundSessionUuid({}))
       );
       const tools = result?.tools ?? [];
       this.cachedTools = tools;
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/daemonMcpProxy.ts]: Fixed in 5764262af. A `discoveryEpoch` (bumped on list_changed and bound-session release) is captured at forward time by listTools/listResources/listResourceTemplates; the response is only stored into the cache when the epoch is unchanged at completion — otherwise it's returned to the caller but the cache stays empty so the next call refetches. Reconnect (`invalidateCache`) does not bump the epoch, so post-reconnect responses still cache. (Per-kind epoch noted as an optional follow-up.)
